File: Design-Patterns-Using-Ruby/Factory/Factory1.rb
# Product interface
class Product
    def operation
        raise NotImplementedError, 'Subclasses must implement the operation method'
    end
end

# Concrete Product classes
class ConcreteProduct1 < Product
    def operation
        puts 'Concrete Product1 Operation!'
    end
end

class ConcreteProduct2 < Product
    def operation
        puts 'Concrete Product2 Operation!'
    end
end

# Creator abstract class
class Factory
    def create_product
        raise NotImplementedError, 'Subclasses must implement the create_product method'
    end
end

# Concrete Creator classes
class ConcreteFactory1 < Factory
    def create_product
        ConcreteProduct1.new
    end
end

class ConcreteFactory2 < Factory
    def create_product
        ConcreteProduct2.new
    end
end

# Client code
factory1 = ConcreteFactory1.new
product1 = factory1.create_product
product1.operation

factory2 = ConcreteFactory2.new
product2 = factory2.create_product
product2.operation
  

=begin 
Concrete Product1 Operation!
Concrete Product2 Operation!
=end

File: Design-Patterns-Using-Go/Factory/Factory1.go
package main

import "fmt"

// Product interface
type Product interface {
	Operation()
}

// Concrete Product classes
type ConcreteProduct1 struct{}

func (cp1 *ConcreteProduct1) Operation() {
	fmt.Println("Concrete Product1 Operation!")
}

type ConcreteProduct2 struct{}

func (cp2 *ConcreteProduct2) Operation() {
	fmt.Println("Concrete Product2 Operation!")
}

// Creator abstract class
type Factory interface {
	CreateProduct() Product
}

// Concrete Creator classes
type ConcreteFactory1 struct{}

func (cf1 *ConcreteFactory1) CreateProduct() Product {
	return &ConcreteProduct1{}
}

type ConcreteFactory2 struct{}

func (cf2 *ConcreteFactory2) CreateProduct() Product {
	return &ConcreteProduct2{}
}

// Client code
func main() {
	f1 := &ConcreteFactory1{}
	p1 := f1.CreateProduct()
	p1.Operation()

	f2 := &ConcreteFactory2{}
	p2 := f2.CreateProduct()
	p2.Operation()
}

/*
Concrete Product1 Operation!
Concrete Product2 Operation!
*/